[iOS] Safari crashes underneath `-[UIPreviewTarget initWithContainer:center:transform:]`

pxlcoder opened the pull request:
#### 418cd9192a514e925cd465b0b0f6aa2b8580e0e6
<pre>
[iOS] Safari crashes underneath `-[UIPreviewTarget initWithContainer:center:transform:]`
<a href="https://bugs.webkit.org/show_bug.cgi?id=300460">https://bugs.webkit.org/show_bug.cgi?id=300460</a>
<a href="https://rdar.apple.com/158751554">rdar://158751554</a>

Reviewed by Abrar Rahman Protyasha, Tim Horton, and Richard Robinson.

Crashes in Safari are being observed due to a UIKit assertion that checks
whether a targeted preview container is in a window.

It remains unclear exactly how WebKit gets into this state. For now,
speculatively fix by not creating a targeted preview when the container is
not in a window. This should have no user observable effect, since the
container is a subview of `_interactionViewsContainerView`, which is a subview
of the `WKWebView`. Consequently, if the container is not in a window, the
web view must also not be.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _createTargetedContextMenuHintPreviewIfPossible]):

Canonical link: <a href="https://commits.webkit.org/301302@main">https://commits.webkit.org/301302@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb7a3e084b4a2bb59139d19f54c74be9793aef5a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125467 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45134 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35877 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132327 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77405 "Built successfully") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45821 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53696 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95545 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9300aa6b-d671-44d6-b742-23c9646f1ef2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128420 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36614 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112203 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76069 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/dd282fd9-0c6e-408d-9549-70b752aab48d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35513 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75801 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106385 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30603 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135005 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52275 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40048 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104022 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52712 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108419 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103769 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49131 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27434 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49430 "Hash fb7a3e08 for PR 52080 does not build (failure)") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19662 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52167 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57952 "") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51518 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54874 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53213 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -11775,17 +11772,21 @@ - (BOOL)positionInformationHasImageOverlayDataDetector
 
 - (UITargetedPreview *)_createTargetedContextMenuHintPreviewIfPossible
 {
+    RetainPtr<UIView> containerForContextMenuHintPreviews = self.containerForContextMenuHintPreviews;
```
The explicit `UIView` is necessary to compile.

aprotyas reviewed (approved):
Maybe a release log when the view has no window?

pxlcoder committed:
[iOS] Safari crashes underneath `-[UIPreviewTarget initWithContainer:center:transform:]`
https://bugs.webkit.org/show_bug.cgi?id=300460
rdar://158751554

Reviewed by Abrar Rahman Protyasha, Tim Horton, and Richard Robinson.

Crashes in Safari are being observed due to a UIKit assertion that checks
whether a targeted preview container is in a window.

It remains unclear exactly how WebKit gets into this state. For now,
speculatively fix by not creating a targeted preview when the container is
not in a window. This should have no user observable effect, since the
container is a subview of `_interactionViewsContainerView`, which is a subview
of the `WKWebView`. Consequently, if the container is not in a window, the
web view must also not be.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _createTargetedContextMenuHintPreviewIfPossible]):

Canonical link: https://commits.webkit.org/301302@main
